text/0724-road-to-typescript.md

```diff
@@ -0,0 +1,236 @@
+
+If Ember officially supports TypeScript, it is important that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present JavaScript and TypeScript (side by side? toggle-able? etc.), where to introduce discussion of TypeScript support in the guides, how API docs should be presented (and whether it may make sense to switch to something like API Extractor) and so on.
+
+The official RFC templates for new features and for deprecations will also need to be updated:
+
+- They must include a description of how changes affect TypeScript consumers.
+- They will need to formally adopt TypeScript types as normative and required for new APIs, formalizing the rough pattern already common for new API designs.
+
+
+##### Authoring and Build
+
+- **How do consumers generate TypeScript apps and addons (and blueprints more generally)?** Should we support `ember new --typescript`? How should blueprints work? This will need to identify the path forward for authoring and publishing, Ember CLI integration, and in particular integration with the Embroider v2 Package Format. (If possible, TypeScript integration should piggy-back on the Embroider v2 *Authoring* Format, but if that format is not available, this is not a blocker.)
+
+- **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider, preferably with the v2 Package Format as the *only* supported publication format. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript: for example, in the decorators implementations, use of the `isolatedModules` setting and associated lints, etc.
+
+- **Migration from the status quo.**  Today, users install `ember-cli-typescript` to opt into using TypeScript, and there is a wholly separate (and, unfortunately, very poorly-maintained) set of blueprints for working with TypeScript.  We will need to identify a migration story from the *current* paired design of `ember-cli-babel` and `ember-cli-typescript` to the future approach suggested by the previous bullet.
+
+
+#### Implementation Details
+
+There are also a number of key implementation concerns which must be addressed, but which, as implementation details, do *not* require RFCs:
+
+- **How are the published types generated?** While much of Ember is implemented in TypeScript, its internal types are not currently ready to be used for public API, and it will likely take some time for them to be ready (with different packages ready at different times). We will need to determine how to publish hand-authored type-definition files, implementation-derived definition files, and over time a mix of the two until we hopefully are able to publish *only* implementation-derived files.
```

Comment:
Would this also want to include the scaffolding effects of generating _types_? Would Ember declare how they fit into the overall application structure? For example:
```
/ app
  / components
  / types
  / routes
```

Reply:
@mwagz I think that would actually fit into the RFC(s) under the [Authoring and Build](https://github.com/chriskrycho/ember-rfcs/blob/intent-to-ts/text/0724-road-to-typescript.md#authoring-and-build) section. And off the top of my head, I don't expect you'd really ever actually *want* to do that particular thing by default, rather than having types live in dedicated source modules, but it could well be that I haven't thought of something—that's what that RFC (or set of RFCs) should cover. Alternatively, depending on the timing of when these RFCs land and when the Embroider v2 Authoring format RFC we need lands, it may make more sense to bundle that as part of that RFC.